File: scala/Problem3.scala
/*
The prime factors of 13195 are 5, 7, 13 and 29.
What is the largest prime factor of the number 600851475143 ?
*/

object Problem3 {
  def factorize(n:Long, d:Long=2, res:List[Long]=Nil):List[Long] = n match {
    case `d` => d::res reverse
    case _ if(n % d == 0) => factorize(n/d, d, d::res)
    case _ => factorize(n, d+1, res)
  }
  def main(args:Array[String]) {
    println(factorize(600851475143L).max)
  }
}

File: perl/Problem3.pl
#!/usr/bin/env perl
# The prime factors of 13195 are 5, 7, 13 and 29.
# 
# What is the largest prime factor of the number 600851475143 ?

use strict;
use warnings;

package Problem3;

sub factor {
    my ($n, $d, @res) = (shift, 2);
    while ($n >= $d) {
        unless ($n % $d) {
            push @res, $d;
            $n /= $d;
        }
        else {
            $d ++;
        }
    }
    return \@res;
}


package main;

print ${Problem3::factor(600851475143)}[-1];

